[JIT] Serialization methods for SourceRange and Source

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22180 [JIT] Load original SourceRanges on import
* #22179 [JIT] Write out SourceRanges on save()
* **#22178 [JIT] Serialization methods for SourceRange and Source**
* #22177 [JIT] Refactor file:line:col to be less ugly

Differential Revision: [D15981426](https://our.internmc.facebook.com/intern/diff/D15981426)

Commit:
[JIT] Serialization methods for SourceRange and Source

Commit:
Update on "[JIT] Serialization methods for SourceRange and Source"

[JIT] Serialization methods for SourceRange and Source

gh-metadata: pytorch pytorch 22178 gh/jamesr66a/7/head

Commit:
Update on "[JIT] Serialization methods for SourceRange and Source"

[JIT] Serialization methods for SourceRange and Source

gh-metadata: pytorch pytorch 22178 gh/jamesr66a/7/head

zdevito commented on torch/csrc/jit/source_range.h:
```diff
@@ -67,6 +68,17 @@ struct Source {
   }
 
+  // Serialize as Tuple[str, Optional[str], int, List[int]]
+  std::shared_ptr<c10::IValue> __getstate__() {
```
shared_ptr<IValue>? that is almost always wrong.

zdevito replied:
Also this code should not be a member of SourceRange or defined in source_range.h it adds a huge dependency of ivalue.h/type.h etc. to source range, which previously included almost no headers. This will dramatically slow build times.

zdevito commented on torch/csrc/jit/source_range.h:
```diff
@@ -78,37 +90,51 @@ struct Source {
+  // serialized IValue representing this Source
+  // Lazily populated, nullptr if not populated
+  //
+  // NB: if you introduce methods that mutate a Source object, you *must*
```
I don't think we can trust that people will read this comment, so it will probably just be a source of bugs. I am not a fan of storing the same information twice in a class for this reason. I think it would be better for this logic to be self-contained in a SourceSerializer object which can manage de-deduping the serialization state.

zdevito reviewed (commented):
The methods need to be put into a separate compilation unit to avoid bloating the dependency tree for source_range objects, which are widely used.

Commit:
Update on "[JIT] Serialization methods for SourceRange and Source"

[JIT] Serialization methods for SourceRange and Source

gh-metadata: pytorch pytorch 22178 gh/jamesr66a/7/head

zdevito commented on torch/csrc/jit/source_range_serializer.h:
```diff
@@ -0,0 +1,39 @@
+  c10::IValue serialize(const SourceRange& sr) {
+    std::vector<c10::IValue> elements = {
+        serialize_source(sr.source()), (int64_t)sr.start(), (int64_t)sr.end()};
+    return c10::ivalue::Tuple::create(elements);
```
nit: move

zdevito commented on torch/csrc/jit/source_range_serializer.h:
```diff
@@ -0,0 +1,39 @@
+    }
+    std::vector<c10::IValue> elements{
+        s->text(), s->filename(), (int64_t)s->starting_line_no()};
+    auto serialized = c10::ivalue::Tuple::create(elements);
```
nit: move

Commit:
Update on "[JIT] Serialization methods for SourceRange and Source"

[JIT] Serialization methods for SourceRange and Source

gh-metadata: pytorch pytorch 22178 gh/jamesr66a/7/head

Commit:
Update on "[JIT] Serialization methods for SourceRange and Source"

[JIT] Serialization methods for SourceRange and Source

gh-metadata: pytorch pytorch 22178 gh/jamesr66a/7/head

Commit:
Update on "[JIT] Serialization methods for SourceRange and Source"

[JIT] Serialization methods for SourceRange and Source

gh-metadata: pytorch pytorch 22178 gh/jamesr66a/7/head
